lzma-rust2/src/lzma_reader.rs

```diff
@@ -199,14 +199,18 @@ impl<R: Read> LZMAReader<R> {
             return Ok(0);
         }
         let mut size = 0;
-        let mut len = buf.len() as u32;
-        let mut off = 0u32;
+        let mut len = buf.len() as u64;
+        let mut off = 0;
         while len > 0 {
             let mut copy_size_max = len;
-            if self.remaining_size <= u64::MAX / 2 && (self.remaining_size as u32) < len {
-                copy_size_max = self.remaining_size as u32;
+            if self.remaining_size <= u64::MAX / 2 && self.remaining_size < len {
+                copy_size_max = self.remaining_size;
+            }
+
+            match usize::try_from(copy_size_max) {
+                Ok(max) => self.lz.set_limit(max),
+                Err(_e) => return Err(std::io::Error::other("lz limit overflow")),
```

Comment:
Two things:
1. I would super lover to recover here, but I don't know how to do that unless we assume the buffer given to `read_decode` is less than `u32::MAX`. I would love to assume that if that's okay.
2. The error message sucks a bit, not sure what to write. Perhaps I could `format!` and stick the `copy_size_max` in there.

Reply:
Oh shoot, I overthought the problem! Sorry about that!

The error case can never happen, even on 32-bit systems. You don't need to use `try_from` and can return to use as simple cast.

Here is why:

`copy_size_max` is either at most `buf.len()`, which we know is of usize.
Or `copy_size_max` is set to `self.remaining_size`, but this is only the case, when we know that `self.remaining_size < len`, which means that it's inside the usize range too.